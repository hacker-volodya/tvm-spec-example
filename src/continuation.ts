import { Cell, Hashmap, Slice } from "ton3-core";
import { OpcodeParser, VarMap } from "./disasm";
import { Instruction } from "./gen/tvm-spec";
import { Stack, StackUnderflowError, StackVariable } from "./stackAnalysis";
import { bitsToIntUint } from "ton3-core/dist/utils/numbers";

const CONTINUATIONS: { [key: string]: string[] } = {
    CALLREF: ["c"],
    JMPREF: ["c"],
    JMPREFDATA: ["c"],
    IFREF: ["c"],
    IFNOTREF: ["c"],
    IFJMPREF: ["c"],
    IFNOTJMPREF: ["c"],
    IFREFELSE: ["c"],
    IFELSEREF: ["c"],
    IFREFELSEREF: ["c1", "c2"],
    IFBITJMPREF: ["c"],
    IFNBITJMPREF: ["c"],
    PUSHCONT: ["s"],
    PUSHCONT_SHORT: ["s"],
}

export class Continuation {
    private args: StackVariable[];
    private code: DecompiledInstruction[];
    private resultStack: StackVariable[];
    private decompileError: any = null;
    private disassembleError: any = null;
    private asmTail: DecodedInstruction[];
    private sliceTail: Slice;

    private constructor(code: DecompiledInstruction[], args: StackVariable[], resultStack: StackVariable[], sliceTail: Slice, asmTail: DecodedInstruction[] = [], decompileError: any = null, disassembleError: any = null) {
        this.args = args;
        this.code = code;
        this.resultStack = resultStack;
        this.asmTail = asmTail;
        this.decompileError = decompileError;
        this.sliceTail = sliceTail;
        this.disassembleError = disassembleError;
    }

    private static dumpOperandConts(operands: VarMap): string[] { 
        let conts = [];
        for (const operand in operands) {
            let v = operands[operand];
            if (v instanceof Continuation) {
                conts.push(v.dump());
                delete operands[operand];
            }
            if (v instanceof Hashmap) {
                v.forEach((methodId: number, cont: Continuation) => {
                    conts.push(`/* methodId: ${methodId} */ ${cont.dump()}`);
                });
                delete operands[operand];
            }
        }
        return conts;
    }

    public dump(): string {
        const indentString = (str: string, count: number, indent = " ") => {
            indent = indent.repeat(count);
            return str.replace(/^/gm, indent);
        };
        const isStackOp = (ins: DecompiledInstruction) => ["stack_basic", "stack_complex"].includes(ins.spec.doc.category);

        // count variable usages
        let useCount: { [name: string]: number } = {};
        for (let instruction of this.code) {
            for (let input of Object.values(instruction.inputs)) {
                let name = (input as any).var.name;
                useCount[name] = (useCount[name] ?? 0) + 1;
            }
        }

        // determine instructions to inline
        let inlineMap: { [name: string]: string } = {};
        let skip = new Set<number>();

        const formatInstruction = (ins: DecompiledInstruction): string => {
            let operandsCopy = { ...ins.operands };
            let conts = Continuation.dumpOperandConts(operandsCopy);
            let operands = Object.values(operandsCopy).map(x => `${x}`);
            let inputVars = Object.values(ins.inputs).map((input: any) => inlineMap[input.var.name] ?? input.var.name);
            let inputStr = operands.concat(...inputVars).concat(...conts).join(', ');
            return `${ins.spec.mnemonic}(${inputStr})`;
        };

        for (let i = 0; i < this.code.length; i++) {
            let instruction = this.code[i];
            if (isStackOp(instruction)) continue;
            let outputs = Object.values(instruction.outputs).map((o: any) => o.var.name);
            if (outputs.length !== 1) continue;
            let outVar = outputs[0];
            if (useCount[outVar] !== 1) continue;
            let j = -1;
            for (let k = i + 1; k < this.code.length; k++) {
                let inputs = Object.values(this.code[k].inputs);
                if (inputs.some((inp: any) => inp.var.name === outVar)) {
                    j = k;
                    break;
                }
            }
            if (j === i + 1 && !isStackOp(this.code[j])) {
                inlineMap[outVar] = formatInstruction(instruction);
                skip.add(i);
            }
        }

        let code = "";
        for (let [index, instruction] of this.code.entries()) {
            if (skip.has(index)) continue;
            // hide stack operations
            if (isStackOp(instruction)) {
                //code += `// ${instruction.spec.mnemonic} ${Object.values(instruction.operands).map(x => `${x}`)}\n`;
                continue;
            }
            let outputVars = Object.values(instruction.outputs).map((output: any) => output.var.name).join(', ');
            let inputVars = Object.values(instruction.inputs).map((input: any) => inlineMap[input.var.name] ?? input.var.name);

            let operandsCopy = { ...instruction.operands };
            let conts = Continuation.dumpOperandConts(operandsCopy);
            let operands = Object.values(operandsCopy).map(x => `${x}`);
            let inputStr = operands.concat(...inputVars).concat(...conts).join(', ');
            code += (outputVars ? `${outputVars} = ` : '') + `${instruction.spec.mnemonic}(${inputStr});\n`;
        }
        code += "// result stack: " + this.resultStack.map(s => s.name).join(', ') + "\n";
        if (this.decompileError) {
            code += "// decompilation error: " + this.decompileError + "\n";
        }
        for (let instruction of this.asmTail) {
            let conts = Continuation.dumpOperandConts(instruction.operands);
            let operands = Object.values(instruction.operands).map(x => `${x}`);
            code += `${instruction.spec.mnemonic} ${operands.concat(...conts).join(', ')}\n`;
        }
        if (this.disassembleError) {
            code += `// Disassemble error: ${this.disassembleError}\n`;
        }
        if (this.sliceTail.bits.length > 0 || this.sliceTail.refs.length > 0) {
            code += `// Tail slice:\n${this.sliceTail}\n`;
        }
        return `function (${this.args.map(a => a.name).join(', ')}) {\n${indentString(code.trimEnd(), 4)}\n}`;
    }

    private static loadOpcode(slice: Slice): DecodedInstruction {
        let [instruction, operands] = OpcodeParser.nextInstruction(slice);
        return {
            spec: instruction,
            operands
        };
    }

    private static decompileInstruction(instruction: DecodedInstruction, stack: Stack): DecompiledInstruction {
        // Handle stack instructions
        if (["stack_basic", "stack_complex"].includes(instruction.spec.doc.category)) {
            stack.execStackInstruction(instruction.spec, instruction.operands);
            return {
                spec: instruction.spec,
                operands: instruction.operands,
                inputs: {},
                outputs: {},
                resultStack: stack.copyEntries()
            };
        }
        // Stop decompilation in case of unknown value flow
        if (instruction.spec.value_flow == undefined || instruction.spec.value_flow.inputs == undefined || instruction.spec.value_flow.outputs == undefined) {
            throw new Error(`instruction is missing value flow: ${instruction.spec.mnemonic}`);
        }
        // Pop inputs
        let stackInputs: VarMap = {};
        if (instruction.spec.value_flow.inputs.stack == undefined) {
            throw new Error(`Unconstrained stack input while parsing ${instruction.spec.mnemonic}`);
        }
        for (let input of instruction.spec.value_flow.inputs.stack.reverse()) {
            if (input.type == 'simple') {
                stackInputs[input.name] = { var: stack.pop(), types: input.value_types };
            } else {
                throw new Error(`not supported stack input '${input.type}' while parsing ${instruction.spec.mnemonic}`);
            }
        }
        // Pop outputs
        let stackOutputs: VarMap = {};
        let constCounter = 0;
        if (instruction.spec.value_flow.outputs.stack == undefined) {
            throw new Error(`Unconstrained stack output while parsing ${instruction.spec.mnemonic}`);
        }
        for (let output of instruction.spec.value_flow.outputs.stack.reverse()) {
            if (output.type == 'simple') {
                stackOutputs[output.name] = { var: stack.push(), types: output.value_types };
            } else if (output.type == 'const') {
                stackOutputs[`const${constCounter++}`] = { var: stack.push(), types: [output.value_type] };
            } else {
                throw new Error(`not supported stack output '${output.type}' while parsing ${instruction.spec.mnemonic}`);
            }
        }
        return {
            spec: instruction.spec,
            operands: instruction.operands,
            inputs: stackInputs,
            outputs: stackOutputs,
            resultStack: stack.copyEntries()
        };
    }

    public static decompile(slice: Slice, initialStack: StackVariable[] = []): Continuation {
        let code: DecompiledInstruction[] = [];
        let stack = new Stack(initialStack);
        let args: StackVariable[] = [];
        let decompileError: any = null;
        let asmTail: DecodedInstruction[] = [];
        while (slice.bits.length > 0) {
            let decodedInsn;
            try {
                decodedInsn = Continuation.loadOpcode(slice);
                if (decodedInsn.spec.mnemonic == "DICTPUSHCONST" && decodedInsn.operands["n"] == 19) {
                    decodedInsn.operands["d"] = Hashmap.parse<Number, Continuation>(19, decodedInsn.operands["d"], {
                        deserializers: {
                            key: k => bitsToIntUint(k, { type: "int" }),
                            value: v => Continuation.decompile(v.slice())
                        }
                    });
                }
                if (CONTINUATIONS[decodedInsn.spec.mnemonic] != undefined) {
                    for (let operand of CONTINUATIONS[decodedInsn.spec.mnemonic]) {
                        decodedInsn.operands[operand] = Continuation.decompile(decodedInsn.operands[operand]);
                    }
                }
            } catch (e) {
                return new Continuation(code, args, stack.copyEntries(), slice, asmTail, decompileError, e);
            }
            if (decompileError == null) {
                for (let t = 0;; t++) {
                    try {
                        console.log("in", stack.dump());
                        console.log(decodedInsn.spec.mnemonic);
                        let stack2 = stack.copy();
                        code.push(Continuation.decompileInstruction(decodedInsn, stack2));
                        stack = stack2;
                        console.log("out", stack.dump());
                        break;
                    } catch (e) {
                        // Check for stack underflow
                        if (e instanceof StackUnderflowError && t < 10) {
                            let newArgs = stack.insertArgs(args.length, e.underflowDepth);
                            args.unshift(...newArgs);
                            continue;
                        }
                        // Stop decompiling, try disassemble tail
                        decompileError = e;
                        break;
                    }
                }
            }
            if (decompileError != null) {
                asmTail.push(decodedInsn);
            }
            // indirect ref jump
            while (slice.bits.length == 0 && slice.refs.length > 0) {
                slice = slice.loadRef().slice();
            }
        }
        return new Continuation(code, args, stack.copyEntries(), slice, asmTail, decompileError);
    }
}

export interface DecodedInstruction {
    spec: Instruction;
    operands: VarMap;
}

export interface DecompiledInstruction {
    spec: Instruction;
    operands: VarMap;
    inputs: VarMap;
    outputs: VarMap;
    resultStack: StackVariable[];
}