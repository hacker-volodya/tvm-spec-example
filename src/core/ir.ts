import type { Instruction, PossibleValueTypes } from "../gen/tvm-spec";
import type { Slice, Cell } from "ton3-core";

// Intermediate Representation (IR) core types

export type IRType = PossibleValueTypes[number];

export type IRValueId = string;

export type IRValueRef = {
  id: IRValueId;
  types?: IRType[];
  continuationMeta?: {
      continuation: IRFunction;
  };
};

export type IRValueDef = {
  id: IRValueId;
  types?: IRType[];
};

// Algebraic data type for instruction operands
export type IROperandValue =
  | { kind: 'int'; value: number }
  | { kind: 'bigint'; value: bigint }
  | { kind: 'bool'; value: boolean }
  | { kind: 'slice'; value: Slice }
  | { kind: 'cell'; value: Cell }
  | { kind: 'cont'; value: IRFunction }
  | { kind: 'cont_map'; value: Map<number, IRFunction> }
  | { kind: 'other'; value: unknown };

export type IROperands = Array<{ name: string; value: IROperandValue }>;

export type IRInlineExpr = {
  kind: 'inline';
  op: IROpPrim;
};

export type IRInputArg = IRValueRef | IRInlineExpr;

export type IRInputs = Array<{ name: string; value: IRInputArg }>;

export type IROutputs = Array<{ name: string; value: IRValueDef }>;

export type IROpPrim = {
  kind: 'prim';
  spec: Instruction;
  mnemonic: string;
  inputs: IRInputs;
  operands: IROperands;
  outputs: IROutputs;
};

export type IRStmt = IROpPrim;

export type IRFunction = {
  kind: 'function';
  name?: string;
  args: IRValueDef[];
  body: IRStmt[];
  result: IRValueRef[];
  asmTail?: { spec: Instruction; operands: IROperands }[];
  tailSliceInfo?: string;
  decompileError?: string | null;
  disassembleError?: string | null;
};

export function isIRFunction(x: unknown): x is IRFunction {
  return !!x && typeof x === 'object' && (x as any).kind === 'function';
}
