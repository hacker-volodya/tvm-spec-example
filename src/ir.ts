import type { Instruction, PossibleValueTypes } from "./gen/tvm-spec";
import type { Slice, Cell, Hashmap } from "ton3-core";

// Intermediate Representation (IR) to describe decompiled code in a
// transformation-friendly way. Optimizers can operate on this IR and
// a later pretty-printer can render textual output.

export type IRType = PossibleValueTypes[number]; // e.g. 'Integer' | 'Cell' | ...

export type IRValueId = string; // uses existing stack variable ids like 'arg0', 'var3'

export type IRValueRef = {
  id: IRValueId;
  types?: IRType[]; // type hints from spec
};

export type IRValueDef = {
  id: IRValueId;
  types?: IRType[];
};

export type IROperandValue = number | bigint | boolean | Slice | Cell | IRFunction | Map<number, IRFunction> | unknown;

export type IROperands = { [name: string]: IROperandValue };

// Input argument can be a reference to a previously defined value
// or an inline expression (used by optimizations such as inlining).
export type IRInlineExpr = {
  kind: 'inline';
  op: IROpPrim; // inlined op; must be side-effect-free to stay correct
};

export type IRInputArg = IRValueRef | IRInlineExpr;

export type IRInputs = { [name: string]: IRInputArg };

export type IROutputs = { [name: string]: IRValueDef };

export type IROpPrim = {
  kind: 'prim';
  spec: Instruction; // full instruction spec for semantics
  mnemonic: string;   // cached for convenience
  inputs: IRInputs;   // stack inputs consumed by this op
  operands: IROperands; // immediate operands (non-stack)
  outputs: IROutputs; // stack outputs produced by this op
};

export type IRStmt = IROpPrim;

export type IRFunction = {
  kind: 'function';
  // Optional human-friendly name for pretty-printer
  name?: string;
  // Formal parameters (top-of-stack last as in decompiler args order)
  args: IRValueDef[];
  // Linear body (we can attach CFG later if needed)
  body: IRStmt[];
  // Resulting stack values after executing the body
  result: IRValueRef[];
  // Optional tails and errors preserved for diagnostics
  asmTail?: { spec: Instruction; operands: IROperands }[];
  tailSliceInfo?: string; // textual form of tail slice for debug only
  decompileError?: string | null;
  disassembleError?: string | null;
};

export function isIRFunction(x: unknown): x is IRFunction {
  return !!x && typeof x === 'object' && (x as any).kind === 'function';
}

// Minimal IR pretty-printer for debugging and evaluation.
export function formatIR(fn: IRFunction, opts?: { methodId?: number }): string {
  const indent = (s: string, n: number) => s.replace(/^/gm, ' '.repeat(n));
  const fmtTypes = (t?: IRType[]) => t && t.length ? `: ${t.join('|')}` : '';
  const fmtValRef = (v: IRValueRef) => `${v.id}${fmtTypes(v.types)}`;
  const fmtValDef = (v: IRValueDef) => `${v.id}${fmtTypes(v.types)}`;

  // Inlined formatter for continuation operands â€“ full body (debug-friendly)
  const formatInlineFn = (f: IRFunction): string => {
    return formatIR(f);
  };

  const formatInlineMap = (m: Map<number, IRFunction>): string => {
    const entries = [...m.entries()];
    const parts = entries.map(([k, fn]) => `${k}: ${formatInlineFn(fn)}`);
    return `{ ${parts.join(', ')} }`;
  };

  const formatInlineOperand = (v: IROperandValue): string => {
    if (isIRFunction(v)) return formatInlineFn(v);
    if (v instanceof Map) return formatInlineMap(v as Map<number, IRFunction>);
    if (typeof v === 'object' && v !== null) {
      // Slice/Cell or other object: fall back to toString()
      try { return String(v); } catch { return '[object]'; }
    }
    if (typeof v === 'number') return formatNumber(v);
    if (typeof v === 'bigint') return formatBigInt(v);
    return JSON.stringify(v);
  };

  const formatNumber = (n: number): string => {
    // Keep small structural constants as decimals to avoid clutter
    if (Number.isInteger(n) && Math.abs(n) <= 512) return String(n);
    // Otherwise include hex for readability
    const hex = '0x' + (n >>> 0).toString(16);
    return `${n} (${hex})`;
  };

  const formatBigInt = (n: bigint): string => {
    // Show bigint in decimal; append hex if small enough
    const abs = n < 0n ? -n : n;
    if (abs <= 512n) return n.toString();
    const hex = '0x' + abs.toString(16);
    return n.toString() + ' (' + hex + ')';
  };

  const formatInlineOpAsExpr = (st: IROpPrim): string => {
    const insOrder = Object.entries(st.inputs).map(([k, v]) => `${k}=${formatInputArg(v)}`).join(', ');
    const opPairs = Object.entries(st.operands).map(([k, v]) => [k, formatInlineOperand(v)] as [string, string]);
    const hasMultiline = opPairs.some(([, v]) => v.includes('\n'));

    if (!hasMultiline) {
      const ops = opPairs.map(([k, v]) => `${k}=${v}`).join(', ');
      const parts = [insOrder, ops].filter(Boolean).join(' | ');
      return `${st.mnemonic}(${parts})`;
    } else {
      let out = `${st.mnemonic}(` + '\n';
      const indent8 = '        ';
      if (insOrder) {
        out += indent8 + insOrder + (opPairs.length ? ' |' : '') + '\n';
      }
      for (let i = 0; i < opPairs.length; i++) {
        const [k, v] = opPairs[i];
        const lines = v.split('\n');
        out += indent8 + k + '=' + lines[0] + (i < opPairs.length - 1 ? ',' : '') + '\n';
        for (let j = 1; j < lines.length; j++) {
          out += indent8 + ' '.repeat(k.length + 1) + lines[j] + (i < opPairs.length - 1 && j === lines.length - 1 ? ',' : '') + '\n';
        }
      }
      out += `    )`;
      return out;
    }
  };

  const formatInputArg = (a: IRInputArg): string => {
    if ((a as any).kind === 'inline') {
      return formatInlineOpAsExpr((a as IRInlineExpr).op);
    }
    return fmtValRef(a as IRValueRef);
  };
  const argsStr = fn.args.map(a => fmtValDef(a)).join(', ');
  const nameFromMethodId = (id?: number) => {
    if (id === -1) return 'recv_external';
    if (id === 0) return 'recv_internal';
    return undefined;
  };
  const nameStr = fn.name ? ` ${fn.name}` : (opts?.methodId !== undefined ? (nameFromMethodId(opts.methodId) ? ` ${nameFromMethodId(opts.methodId)}` : '') : '');
  const header = opts?.methodId !== undefined ? `/* methodId: ${opts.methodId} */\n` : '';
  let out = header + `function${nameStr} (${argsStr}) {\n`;
  for (const st of fn.body) {
    // Order outputs according to spec value_flow if available
    const getOutputsInOrder = (): IRValueDef[] => {
      const vf: any = (st as any).spec?.value_flow;
      const specStack = vf?.outputs?.stack as any[] | undefined;
      if (!specStack) return Object.values(st.outputs);
      const entries = Object.entries(st.outputs) as [string, IRValueDef][];
      const byName = new Map(entries);
      const constKeys: string[] = entries.filter(([k]) => k.startsWith('const')).map(([k]) => k);
      const used = new Set<string>();
      const ordered: IRValueDef[] = [];
      for (const out of specStack) {
        if (out.type === 'simple') {
          const k = out.name as string;
          const v = byName.get(k);
          if (v) { ordered.push(v as IRValueDef); used.add(k); }
        } else if (out.type === 'const') {
          // consume next const* entry in insertion order
          const k = constKeys.find((x) => !used.has(x));
          if (k) { ordered.push(byName.get(k)!); used.add(k); }
        }
        // conditional outputs are not represented in st.outputs
      }
      // Append any remaining outputs in original insertion order
      for (const [k, v] of entries) {
        if (!used.has(k)) ordered.push(v);
      }
      return ordered;
    };

    const outs = getOutputsInOrder().map(fmtValDef).join(', ');
    const insOrder = Object.entries(st.inputs).map(([k, v]) => `${k}=${formatInputArg(v)}`).join(', ');
    const opPairs = Object.entries(st.operands).map(([k, v]) => [k, formatInlineOperand(v)] as [string, string]);
    const hasMultiline = opPairs.some(([, v]) => v.includes('\n'));

    if (!hasMultiline) {
      const ops = opPairs.map(([k, v]) => `${k}=${v}`).join(', ');
      const parts = [insOrder, ops].filter(Boolean).join(' | ');
      out += `    ${outs ? outs + ' = ' : ''}${st.mnemonic}(${parts});\n`;
    } else {
      out += `    ${outs ? outs + ' = ' : ''}${st.mnemonic}(` + '\n';
      const indent8 = '        ';
      if (insOrder) {
        out += indent8 + insOrder + (opPairs.length ? ' |' : '') + '\n';
      }
      for (let i = 0; i < opPairs.length; i++) {
        const [k, v] = opPairs[i];
        const lines = v.split('\n');
        out += indent8 + k + '=' + lines[0] + (i < opPairs.length - 1 ? ',' : '') + '\n';
        for (let j = 1; j < lines.length; j++) {
          out += indent8 + ' '.repeat(k.length + 1) + lines[j] + (i < opPairs.length - 1 && j === lines.length - 1 ? ',' : '') + '\n';
        }
      }
      out += `    );\n`;
    }
  }
  if (fn.result.length) {
    out += `    // result: ${fn.result.map(fmtValRef).join(', ')}\n`;
  }
  if (fn.decompileError) out += `    // decompilation error: ${fn.decompileError}\n`;
  if (fn.asmTail && fn.asmTail.length) {
    for (const ins of fn.asmTail) {
      const ops = Object.entries(ins.operands).map(([k, v]) => `${k}=${formatInlineOperand(v)}`).join(', ');
      out += `    ${ins.spec.mnemonic} ${ops}\n`;
    }
  }
  if (fn.disassembleError) out += `    // disassemble error: ${fn.disassembleError}\n`;
  if (fn.tailSliceInfo) {
    out += `    // tail slice: ${fn.tailSliceInfo}\n`;
  }
  out += `}`;
  return out;
}
