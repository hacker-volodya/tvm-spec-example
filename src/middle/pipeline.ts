import type { IRFunction } from "../ir";
import { inlinePrevSingleUse, inlineSingleUseConsts } from "../opt/inline";

export type Pass = (fn: IRFunction) => IRFunction;

export class Pipeline {
  private passes: Pass[] = [];

  use(pass: Pass): this {
    this.passes.push(pass);
    return this;
  }

  run(fn: IRFunction): IRFunction {
    return this.passes.reduce((acc, p) => p(acc), fn);
  }
}

export function defaultPipeline(): Pipeline {
  return new Pipeline()
    .use(inlineSingleUseConsts)
    .use(inlinePrevSingleUse);
}

