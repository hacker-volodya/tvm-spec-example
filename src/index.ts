import { BOC, Hashmap, Builder, Slice } from "ton3-core"
import * as fs from 'fs'
import { Continuation } from "./continuation";
import { formatIR } from "./ir";
import { inlinePrevSingleUse } from "./opt/inline";
import { OpcodeParser } from "./disasm";
import { bitsToIntUint } from "ton3-core/dist/utils/numbers";


const slice = BOC.from(new Uint8Array(fs.readFileSync(process.argv[2]))).root[0].slice();

function cloneSlice(s: Slice): Slice {
    return new Builder().storeBits(s.bits).storeRefs(s.refs).cell().slice();
}

function tryDecodeFunctionDictFromRoot(root: Slice): Map<number, ReturnType<Continuation['toIR']>> | null {
    const sc = cloneSlice(root);
    try {
        // 1) SETCP
        const [ins1] = OpcodeParser.nextInstruction(sc);
        if (ins1.mnemonic !== 'SETCP') return null;
        // 2) DICTPUSHCONST
        const [ins2, ops2] = OpcodeParser.nextInstruction(sc);
        if (ins2.mnemonic !== 'DICTPUSHCONST') return null;
        const n: number = ops2['n'];
        const d: Slice = ops2['d'];
        // 3) DICTIGETJMPZ
        const [ins3] = OpcodeParser.nextInstruction(sc);
        if (ins3.mnemonic !== 'DICTIGETJMPZ') return null;
        // 4) THROWARG
        const [ins4] = OpcodeParser.nextInstruction(sc);
        if (ins4.mnemonic !== 'THROWARG') return null;
        // ensure no extra code in root
        if (sc.bits.length !== 0 || sc.refs.length !== 0) return null;

        // Parse dictionary as functions: key=int, value=Continuation
        const dict = Hashmap.parse<number, ReturnType<typeof Continuation.decompile>>(n, d, {
            deserializers: {
                key: (k: any) => bitsToIntUint(k, { type: 'int' }) as number,
                value: (v: any) => Continuation.decompile(v.slice())
            }
        });

        const result = new Map<number, ReturnType<Continuation['toIR']>>();
        dict.forEach((methodId: number, cont: ReturnType<typeof Continuation.decompile>) => {
            result.set(methodId, cont.toIR());
        });
        return result;
    } catch (_) {
        return null;
    }
}

const dictIR = tryDecodeFunctionDictFromRoot(slice);
if (dictIR) {
    // Print dictionary of functions, sorted by methodId
    const pairs = Array.from(dictIR.entries()).sort((a, b) => a[0] - b[0]);
    for (const [id, fn] of pairs) {
        console.log(`/* methodId: ${id} */`);
        console.log(formatIR(inlinePrevSingleUse(fn)));
        console.log();
    }
} else {
    const cont = Continuation.decompile(slice);
    console.log(formatIR(inlinePrevSingleUse(cont.toIR())))
}
