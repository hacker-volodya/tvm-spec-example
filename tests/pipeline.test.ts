import { defaultPipeline } from '../src/middle/pipeline';
import { IRFunction, IROpPrim, IRValueDef, IRValueRef } from '../src/ir';
import type { Instruction } from '../src/gen/tvm-spec';

function fakeSpec(category: string): Instruction {
  // Provide only fields used by the code under test; cast to Instruction
  const spec: any = {
    mnemonic: 'FAKE',
    since_version: 9999,
    doc: {
      opcode: '',
      stack: '',
      category,
      description: '',
      gas: '',
      fift: '',
      fift_examples: [],
    },
    bytecode: { tlb: '', prefix: '00', operands: [] },
    value_flow: { inputs: { registers: [] }, outputs: { registers: [] } },
    control_flow: { branches: [], nobranch: true },
  };
  return spec as Instruction;
}

function op(
  mnemonic: string,
  outputs: Record<string, IRValueDef>,
  inputs: Record<string, IRValueRef | { kind: 'inline'; op: IROpPrim }> = {},
  category = 'misc'
): IROpPrim {
  return {
    kind: 'prim',
    spec: fakeSpec(category),
    mnemonic,
    inputs,
    operands: {},
    outputs,
  };
}

describe('defaultPipeline snapshot', () => {
  test('inlines single-use const producer into consumer', () => {
    const v1: IRValueDef = { id: 'v1' };
    const y1: IRValueDef = { id: 'y1' };

    const constOp = op('PUSHCONST', { v: v1 }, {}, 'const_int');
    const consumer = op('ADD', { y: y1 }, { a: { id: 'v1' }, b: { id: 'arg0' } });

    const fn: IRFunction = {
      kind: 'function',
      args: [{ id: 'arg0' }],
      body: [constOp, consumer],
      result: [{ id: 'y1' }],
    };

    const out = defaultPipeline().run(fn);
    expect(out).toMatchSnapshot();
  });

  test('inlines previous single-use non-const op', () => {
    const t: IRValueDef = { id: 't' };
    const y: IRValueDef = { id: 'y' };

    const prev = op('SHL', { t }, { a: { id: 'arg0' } }, 'misc');
    const curr = op('NEGATE', { y }, { x: { id: 't' } }, 'misc');

    const fn: IRFunction = {
      kind: 'function',
      args: [{ id: 'arg0' }],
      body: [prev, curr],
      result: [{ id: 'y' }],
    };

    const out = defaultPipeline().run(fn);
    expect(out).toMatchSnapshot();
  });

  test('does not inline producer if its value is part of result', () => {
    const v: IRValueDef = { id: 'vres' };
    const y: IRValueDef = { id: 'y' };

    const producer = op('PUSHCONST', { v }, {}, 'const_int');
    // Single use in consumer, but value is also listed in result
    const consumer = op('USE', { y }, { a: { id: 'vres' } }, 'misc');

    const fn: IRFunction = {
      kind: 'function',
      args: [],
      body: [producer, consumer],
      result: [{ id: 'vres' }, { id: 'y' }],
    };

    const out = defaultPipeline().run(fn);
    expect(out).toMatchSnapshot();
  });
});
