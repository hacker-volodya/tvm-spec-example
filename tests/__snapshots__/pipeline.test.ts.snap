// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defaultPipeline snapshot does not inline producer if its value is part of result 1`] = `
{
  "args": [],
  "body": [
    {
      "inputs": {},
      "kind": "prim",
      "mnemonic": "PUSHCONST",
      "operands": {},
      "outputs": {
        "v": {
          "id": "vres",
        },
      },
      "spec": {
        "bytecode": {
          "operands": [],
          "prefix": "00",
          "tlb": "",
        },
        "control_flow": {
          "branches": [],
          "nobranch": true,
        },
        "doc": {
          "category": "const_int",
          "description": "",
          "fift": "",
          "fift_examples": [],
          "gas": "",
          "opcode": "",
          "stack": "",
        },
        "mnemonic": "FAKE",
        "since_version": 9999,
        "value_flow": {
          "inputs": {
            "registers": [],
          },
          "outputs": {
            "registers": [],
          },
        },
      },
    },
    {
      "inputs": {
        "a": {
          "id": "vres",
        },
      },
      "kind": "prim",
      "mnemonic": "USE",
      "operands": {},
      "outputs": {
        "y": {
          "id": "y",
        },
      },
      "spec": {
        "bytecode": {
          "operands": [],
          "prefix": "00",
          "tlb": "",
        },
        "control_flow": {
          "branches": [],
          "nobranch": true,
        },
        "doc": {
          "category": "misc",
          "description": "",
          "fift": "",
          "fift_examples": [],
          "gas": "",
          "opcode": "",
          "stack": "",
        },
        "mnemonic": "FAKE",
        "since_version": 9999,
        "value_flow": {
          "inputs": {
            "registers": [],
          },
          "outputs": {
            "registers": [],
          },
        },
      },
    },
  ],
  "kind": "function",
  "result": [
    {
      "id": "vres",
    },
    {
      "id": "y",
    },
  ],
}
`;

exports[`defaultPipeline snapshot inlines previous single-use non-const op 1`] = `
{
  "args": [
    {
      "id": "arg0",
    },
  ],
  "body": [
    {
      "inputs": {
        "x": {
          "kind": "inline",
          "op": {
            "inputs": {
              "a": {
                "id": "arg0",
              },
            },
            "kind": "prim",
            "mnemonic": "SHL",
            "operands": {},
            "outputs": {
              "t": {
                "id": "t",
              },
            },
            "spec": {
              "bytecode": {
                "operands": [],
                "prefix": "00",
                "tlb": "",
              },
              "control_flow": {
                "branches": [],
                "nobranch": true,
              },
              "doc": {
                "category": "misc",
                "description": "",
                "fift": "",
                "fift_examples": [],
                "gas": "",
                "opcode": "",
                "stack": "",
              },
              "mnemonic": "FAKE",
              "since_version": 9999,
              "value_flow": {
                "inputs": {
                  "registers": [],
                },
                "outputs": {
                  "registers": [],
                },
              },
            },
          },
        },
      },
      "kind": "prim",
      "mnemonic": "NEGATE",
      "operands": {},
      "outputs": {
        "y": {
          "id": "y",
        },
      },
      "spec": {
        "bytecode": {
          "operands": [],
          "prefix": "00",
          "tlb": "",
        },
        "control_flow": {
          "branches": [],
          "nobranch": true,
        },
        "doc": {
          "category": "misc",
          "description": "",
          "fift": "",
          "fift_examples": [],
          "gas": "",
          "opcode": "",
          "stack": "",
        },
        "mnemonic": "FAKE",
        "since_version": 9999,
        "value_flow": {
          "inputs": {
            "registers": [],
          },
          "outputs": {
            "registers": [],
          },
        },
      },
    },
  ],
  "kind": "function",
  "result": [
    {
      "id": "y",
    },
  ],
}
`;

exports[`defaultPipeline snapshot inlines single-use const producer into consumer 1`] = `
{
  "args": [
    {
      "id": "arg0",
    },
  ],
  "body": [
    {
      "inputs": {
        "a": {
          "kind": "inline",
          "op": {
            "inputs": {},
            "kind": "prim",
            "mnemonic": "PUSHCONST",
            "operands": {},
            "outputs": {
              "v": {
                "id": "v1",
              },
            },
            "spec": {
              "bytecode": {
                "operands": [],
                "prefix": "00",
                "tlb": "",
              },
              "control_flow": {
                "branches": [],
                "nobranch": true,
              },
              "doc": {
                "category": "const_int",
                "description": "",
                "fift": "",
                "fift_examples": [],
                "gas": "",
                "opcode": "",
                "stack": "",
              },
              "mnemonic": "FAKE",
              "since_version": 9999,
              "value_flow": {
                "inputs": {
                  "registers": [],
                },
                "outputs": {
                  "registers": [],
                },
              },
            },
          },
        },
        "b": {
          "id": "arg0",
        },
      },
      "kind": "prim",
      "mnemonic": "ADD",
      "operands": {},
      "outputs": {
        "y": {
          "id": "y1",
        },
      },
      "spec": {
        "bytecode": {
          "operands": [],
          "prefix": "00",
          "tlb": "",
        },
        "control_flow": {
          "branches": [],
          "nobranch": true,
        },
        "doc": {
          "category": "misc",
          "description": "",
          "fift": "",
          "fift_examples": [],
          "gas": "",
          "opcode": "",
          "stack": "",
        },
        "mnemonic": "FAKE",
        "since_version": 9999,
        "value_flow": {
          "inputs": {
            "registers": [],
          },
          "outputs": {
            "registers": [],
          },
        },
      },
    },
  ],
  "kind": "function",
  "result": [
    {
      "id": "y1",
    },
  ],
}
`;
